returnn/tf/layers/basic.py

```diff
@@ -3414,35 +3406,47 @@ def get_out_data_from_opts(cls, name, axis, dims, pad_to_multiples=None, sources
     resolved_dims = None
     if dims and isinstance(dims[0], Dim):
       assert all(isinstance(d, Dim) for d in dims)
-      resolved_dims = dims
+      resolved_dims = tuple(dims)
       dims = [d.dimension or -1 for d in dims]
     if data.batch_shape[axis] is not None:
       resolved_shape_dims = cls._resolve_dims(
         old_dim=data.batch_shape[axis], new_dims=dims, pad_to_multiples=pad_to_multiples)
     else:
       resolved_shape_dims = tuple([(d if (d >= 0) else None) for d in dims])
 
-    rem_dim_indices = [i for i, d in enumerate(dims) if d < 0]
-    assert len(rem_dim_indices) <= 1, "only one entry in dims %r can be -1" % (dims,)
-    rem_dim_idx = rem_dim_indices[0] if rem_dim_indices else (len(dims) - 1)
-    axis_dim_tag = data.dim_tags[axis]
-    if all(d == 1 for (i, d) in enumerate(dims) if i != rem_dim_idx):
-      rem_dim = axis_dim_tag  # we can overtake the existing dim tag
-      if resolved_dims:
-        assert resolved_dims[rem_dim_idx] == axis_dim_tag
-    else:
-      rem_dim = Dim(
-        kind=axis_dim_tag.kind,
-        description="%s_split_dims%i_rem" % (name, rem_dim_idx),
-        dimension=resolved_shape_dims[rem_dim_idx])
     if not resolved_dims:
       resolved_dims = tuple(
         Dim(
-          kind=Dim.Types.Spatial,
+          kind=Dim.Types.Spatial if shape_dim is None else Dim.Types.Feature,
           description="%s_split_dims%i" % (name, i),
-          dimension=resolved_shape_dims[i])
-        if i != rem_dim_idx else rem_dim
-        for i in range(len(dims)))
+          dimension=shape_dim)
+        for i, shape_dim in enumerate(resolved_shape_dims))
+    rem_dim_indices = [i for i, d in enumerate(dims) if d < 0]
+    assert len(rem_dim_indices) <= 1, "only one entry in dims %r can be -1" % (dims,)
+    if len(rem_dim_indices) == 1:
+      rem_dim_idx = rem_dim_indices[0]
+      rem_dim = resolved_dims[rem_dim_idx]
+      axis_dim_tag = data.dim_tags[axis]
+      import numpy
+      n = int(numpy.prod([dim for dim in dims if dim > 0]))
+      if rem_dim.batch is None:
+        rem_dim.batch = axis_dim_tag.batch
+      if rem_dim.dimension is None:  # need to set dyn_size
+        if rem_dim.dyn_size_ext is not None:
+          pass  # assume it is set up correctly
+        if axis_dim_tag.dimension is not None:  # out dim stays static if in dim was static
+          assert axis_dim_tag.dimension % n == 0
+          rem_dim.dimension = axis_dim_tag.dimension // n
+        elif n == 1:  # can overtake old dyn size
+          assert axis_dim_tag.dyn_size_ext is not None
+          rem_dim.dyn_size_ext = axis_dim_tag.dyn_size_ext
+          rem_dim.derived_from_tag = axis_dim_tag
```

Comment:
`derived_from_tag` is used incorrectly. It means that it is derived from some other tag, but different to the other tag.

But here we have the case that it is the same. So you could use `declare_same_as`. Or why not directly?
```
rem_dim = axis_dim_tag
```

As it was also in the old code.

Reply:
This is not needed if the user doesn't specify dims as a tag list. I will change it so that then it actually just re-uses the dim tag of the input.
However, if you e.g. use `dims = [SpatialDim("rem"), FeatureDim("broadcast", 1)]` or so, like I have it in my test, then I need to properly assign `dyn_size_ext` to the new `rem` dim tag.

Reply:
In that case, should the user-specified "rem" dim tag set derived_from to the input axis?

Reply:
No, if there is an existing object already, you should use `declare_same_as`. I think we have some similar logic in other layers.

Reply:
For this case where it is the same. If it is not the same (because some other dims > 1), then you can set `derived_from`.